src/performance.js

```diff
@@ -58,8 +66,48 @@ export class Performance {
 
     /** @const @private {!Array<TickEventDef>} */
     this.events_ = [];
+
+    /** @private {boolean} */
+    this.isViewerVisible_ = false;
+
+    this.win.addEventListener('unload', this.flush.bind(this));
   }
 
+  /**
+   * Gets primary amp core services.
+   * @return {!Performance}
+   */
+  init() {
+    this.viewer_ = viewerFor(this.win);
+    this.resources_ = resourcesFor(this.win);
+    return this;
+  }
+
+  /**
+   * Listens to viewer and resource events.
+   */
+  listen() {
+    if (!(this.viewer_ && this.resources_)) {
+      return;
+    }
+
+    // User requested document to be visible
+    this.viewer_.whenFirstVisible().then(() => {
+      this.isViewerVisible_ = true;
+      this.tick('rdc');
+    });
+
+    this.resources_.whenFirstResourceLaidOut().then(() => {
+      if (this.isViewerVisible_) {
+        // Prerender Completed, but user already
```

Comment:
No, not necessarily. We need to go back and discuss what we want the event `pc` to be exactly. In this block of code there's only one thing you know: if `!isViewerVisible` it means that the layout has happened either in prerender phase. To know more - we'd probably need to add more things to viewer, such as count number of times visibility state changed or such. Or we can also expand our protocol with the viewer so that the viewer can tell AMP document things like "please prerender now" vs "no longer visible". That being said, I think we can safely assume that when document is opened in "hidden" state - we need to prerender and when document goes from visible to hidden - it's being discarded.